src/rust/ide/ast/impl/src/known.rs

```diff
@@ -58,23 +46,48 @@ impl<T> KnownAst<T> {
 
     /// Returns a reference to the stored `Ast` with `Shape` of `T`.
     pub fn ast(&self) -> &Ast { &self.ast }
+}
+
+impl<T,E> KnownAst<T>
+where for<'t> &'t Shape<Ast> : TryInto<&'t T,Error=E> {
+
+    /// Checks if the shape of given Ast node is compatible with `T`.
+    /// If yes, returns Ok with Ast node wrapped as KnownAst.
+    /// Otherwise, returns an error.
+    pub fn try_new(ast:Ast) -> Result<KnownAst<T>,E> {
+        if let Some(error_matching) = ast.shape().try_into().err() {
+            Err(error_matching)
+        } else {
+            Ok(KnownAst {ast,phantom:default()})
+        }
+    }
 
     /// Returns the AST's shape.
-    pub fn shape<E>(&self) -> &T
-    where for<'t> &'t Shape<Ast> : TryInto<&'t T,Error=E>,
-          E                      : Debug, {
+    pub fn shape(&self) -> &T
+    where E : Debug {
         self.deref()
     }
 
     /// Updated self in place by applying given function on the stored Shape.
-    pub fn update_shape<E>(&mut self, f:impl FnOnce(&mut T))
-    where for<'t> &'t Shape<Ast> : TryInto<&'t T,Error=E>,
-          T                      : Clone + Into<Shape<Ast>>,
-          E                      : Debug {
+    pub fn update_shape(&mut self, f:impl FnOnce(&mut T))
+    where T : Clone + Into<Shape<Ast>>,
+          E : Debug {
         let mut shape = self.shape().clone();
         f(&mut shape);
         self.ast = self.ast.with_shape(shape)
     }
+
+    /// Create new instance of KnownAst with mapped shape.
+    pub fn map_shape<S,E1>(&self, f:impl FnOnce(T) -> S) -> KnownAst<S>
+    where for<'t> &'t Shape<Ast> : TryInto<&'t S,Error=E1>,
+          T                      : Clone + Into<Shape<Ast>>,
```

Comment:
Dunno, it haven't work without it

Reply:
Adam, could you try it again? It's wrong. `shape` should be clone. If it was inferred to be `&T` maybe you were missing constraint `T:Clone` back then. It should work. Moreover, let's not close such comments by stating "I don't know" - we need to know why and adding constraints because compiler told so is a very bad thing and we will have problems because of that.